### PR TITLE
feat: new availableLocales app config to ease i18n lazy loading

### LINKED ISSUE #2447: Lazy load translations bundles automatically in Custom Applications
**Background**
When working with new Custom Applications, developers are required to provide the translations to the [ApplicationShell](https://docs.commercetools.com/custom-applications/components/application-shell) component. Both an object or a list of promises can be provided, the latter being preferred when the user needs to use several languages so the component can lazy-load only the translations needed for the navigation language.
There is [this documentation](https://docs.commercetools.com/custom-applications/development/translations#using-the-messages-in-the-application) explaining how to correctly generate the promises list but it seems a boilerplate code every Custom Applications developer must write.

**Describe the solution you'd like**
Since the [process of generating the translations bundles](https://docs.commercetools.com/custom-applications/development/translations#extracting-messages-for-translations) is configured to output its result in a predefined and known directory (`/src/i18n/data/*`), the `ApplicationShell` component could have the default behaviour of lazy-load the right translation bundle.
We could keep the [applicationMessages](https://docs.commercetools.com/custom-applications/components/application-shell#properties) property for more elaborated use cases a developer might find.

### PR DESCRIPTION
**Pull request by CarlosCortizasCT**
resolves #2447

TODO:
- [ ] Testing
- [ ] Update public documentation
- [x] Changeset

### CODE REVIEW

**Comment by CarlosCortizasCT** [comment]
I'm having this warning when running the new tests:
image

It's complaining because one of the test paths logs a warning to the console and that's not allowed to happen in the tests but I think this is a fair use case.

I found [this commit](https://github.com/commercetools/merchant-center-application-kit/pull/2254/files) where a helper was added to help with this situations but I don't seem to find the right way to make it work (I tried copy some samples of the frontend repository).

Any help would be very much welcome 😃

**Comment by emmenko** [comment]
If you expect to have something logged, you can mock `console` in the test scenario and assert that it was called with the right message. I think we already have a bunch of cases like this.

File: packages/application-config/src/schema.ts
Comment (emmenko) [reply packages/application-config/src/schema.ts:170]: Also, the supported locales should be listed as enums, so that the config is correctly validated.

As for the name `availableLocales`, I'm not super convinced about this. I'll think about it.

File: playground/src/i18n/data/es.json
Comment (emmenko) [inline playground/src/i18n/data/es.json:13]: Thanks! 😜

**Review by emmenko — CHANGES_REQUESTED** [review]
Thanks, I like where this is going 👌

I left some comments as there are still some little details to figure out, otherwise looks good.

File: packages/application-shell/src/components/application-shell/helpers.ts
Comment (emmenko) [reply packages/application-shell/src/components/application-shell/helpers.ts:45]: I feel like this is the main challenge to figure out. Maybe we can do something like this, using the `webpack.DefinePlugin`

Screen Shot 2022-01-24 at 09 37 44

File: packages/application-config/src/schema.ts
Comment (CarlosCortizasCT) [reply packages/application-config/src/schema.ts:170]: I updated the `schema.json` file with the new property (I'll change its name when the final one it's decided) and the tests for the `process-config` command.

Regarding the _enum_, the locales we will support are all of them, right? I mean, anyone creating a custom app can decide which locales they want to use within their app so I'm not sure how we can validate that with an enum.
Should we list every possible language or maybe we should stick to the ones defined for the `labelAllLocales`?

File: packages/application-shell/src/components/application-shell/helpers.ts
Comment (CarlosCortizasCT) [reply packages/application-shell/src/components/application-shell/helpers.ts]: Updated in [this commit](https://github.com/commercetools/merchant-center-application-kit/pull/2452/commits/7563a424750b5cee04acb48cc29e336d6eccec7b)

File: packages/application-config/src/schema.ts
Comment (emmenko) [reply packages/application-config/src/schema.ts:170]: > Regarding the enum, the locales we will support are all of them, right? 

No, there a specific number of locales that we support. 

```
['en', 'de', 'es', 'fr-FR', 'zh-CN', 'ja']
```

PS: we should probably document this a bit better in the new docs.

File: packages/application-shell/src/components/application-shell/helpers.ts
Comment (CarlosCortizasCT) [reply packages/application-shell/src/components/application-shell/helpers.ts]: Well, I think we have a problem here with my initial expectations because of the way [dynamic expressions](https://webpack.js.org/api/module-methods/#dynamic-expressions-in-import) work in Webpack dynamic imports.

It is not possible to use a complete dynamic path because webpack needs to know about the possible files it would need to lazy load. So you give it a relative path and it creates a bundle for every possible combinations.

As the `ApplicationShell` component is imported from the consumer application, I don't seem to find the right path to use in the different situations (locally development in the app-kit, locally development from the mc-frontent and production both default and custom apps).

The only idea I came up with is to delegate to the consumer app the call to the dynamic `import` so it can correctly set the translations folder and leave it to the `ApplicationShell` to figure out which file should be loaded.

From the `ApplicationShell` perspective, it would look like this:
```
const EntryPoint = () => (
  <ApplicationShell
    environment={window.app}
    localeLoader={(localeFileName) =>
      import(
        `../../i18n/data/${localeFileName}` /* webpackChunkName: "app-i18n-[request]" */
      )
    }
    onRegisterErrorListeners={({ dispatch }) => {
      Sdk.Get.errorHandler = (error) =>
        globalActions.handleActionError(error, 'sdk')(dispatch);
    }}
    render={() => <ApplicationStateMachines />}
  />
);
```

I'm not very proud of this solution, but it's the only I currently have.

Regarding the proposed ```availableLocales``` prop, after thinking about it twice, it doesn't seem too valuable to me.
It is only used to validate the language requested is configured by the developers of Apps, but the component would anyway return the default translations (english) so I don't see the point of having it actually.

I pushed a [commit}(https://github.com/commercetools/merchant-center-application-kit/pull/2452/commits/03f75350531e8c8abf983a5dd85bb5daf5d3e90e) with the implementation for this idea to better explaining but I wouldn't like to go further before getting some feedback about it.

Comment (emmenko) [reply packages/application-shell/src/components/application-shell/helpers.ts]: So it didn't work when using `webpack.DefinePlugin`?

Comment (emmenko) [reply packages/application-shell/src/components/application-shell/helpers.ts]: Ah but I guess the problem is also when we bundle the packages. Rollup will find dynamic imports for code splitting but we actually don't want to execute it just yet.

So yeah, I guess using the `import()` syntax can only be done in the application code itself.

There was actually another small issue that I forgot to mention. The translation files are usually located in `i18n/data`, however we also use the option to get the "compiled" version of the translation object which are located in `i18n/compiled-data`.

At this point I feel like we're going down a bit the rabbit hole, as there are no more clear benefits with a new approach.

Let's have a chat about this but I suppose we should put on hold this for now.

**Comment by CarlosCortizasCT** [comment]
After discussing the current status of this PR with @emmenko and given the problems we discovered with dynamic imports and the lack of ability to actually provide a solution which could add a significant value over the current situation, and also that there is already a discussion opened regarding how we should improve translations management at runtime in the [frontend chapter](https://github.com/commercetools/fe-chapter-notes/issues/286), we decided to ditch this PR and focus in that new approach.

**Pull request closed without merging**
